File: BlinkCard/android/src/main/java/com/microblink/blinkcard/reactnative/serialization/BlinkCardSerializationUtils.kt
package com.microblink.blinkcard.reactnative.serialization

import android.graphics.Bitmap
import android.util.Base64
import com.microblink.blinkcard.core.result.BlinkCardSingleSideScanningResult
import com.microblink.blinkcard.core.result.CardAccountResult
import com.microblink.blinkcard.core.result.CardLivenessCheckResult
import com.microblink.blinkcard.core.result.CroppedImageResult
import com.microblink.blinkcard.core.result.DateResult
import com.microblink.blinkcard.core.session.BlinkCardScanningResult
import org.json.JSONObject
import java.io.ByteArrayOutputStream

object BlinkCardSerializationUtils {
  fun serializeBlinkCardScanningResult(result: BlinkCardScanningResult): String {
    val resultJson: MutableMap<String, Any> = mutableMapOf()

    result.issuingNetwork.let { resultJson["issuingNetwork"] = it }
    result.iban?.let { resultJson["iban"] = it }
    result.cardholderName?.let { resultJson["cardholderName"] = it }
    result.overallCardLivenessResult.let { resultJson["overallCardLivenessResult"] = it.name.lowercase() }
    result.cardAccounts.let {
      resultJson["cardAccounts"] = it.map { cardResult -> serializeCardAccountResult(cardResult) }
    }
    result.firstSideResult?.let { resultJson["firstSideResult"] = serializeSingleSideScanningResult(it) }
    result.secondSideResult?.let { resultJson["secondSideResult"] = serializeSingleSideScanningResult(it) }

    return JSONObject(resultJson).toString()
  }

  private fun serializeCardAccountResult(cardAccountResult: CardAccountResult): Map<String, Any> {
    val resultJson: MutableMap<String, Any> = mutableMapOf(
      "cardNumber" to cardAccountResult.cardNumber,
      "cardNumberValid" to cardAccountResult.cardNumberValid,
    )
    cardAccountResult.cardNumberPrefix?.let { resultJson["cardNumberPrefix"] = it }
    cardAccountResult.cvv?.let { resultJson["cvv"] = it }
    cardAccountResult.expiryDate?.let { resultJson["expiryDate"] = serializeDateResult(it) }
    cardAccountResult.cardCategory?.let { resultJson["cardCategory"] = it }
    cardAccountResult.issuerName?.let { resultJson["issuerName"] = it }
    cardAccountResult.issuerCountryCode?.let { resultJson["issuerCountryCode"] = it }
    cardAccountResult.issuerCountry?.let { resultJson["issuerCountry"] = it }
    cardAccountResult.issuerName?.let { resultJson["issuerName"] = it }
    cardAccountResult.issuerName?.let { resultJson["issuerName"] = it }

    return resultJson
  }

  private fun <T> serializeDateResult(dateResult: DateResult<T>): Map<String, Any> {
    val resultJson: MutableMap<String, Any> = mutableMapOf()

    dateResult.day?.let { resultJson["day"] = it }
    dateResult.month?.let { resultJson["month"] = it }
    dateResult.year?.let { resultJson["year"] = it }
    dateResult.originalString?.let { resultJson["originalString"] = it }
    dateResult.filledByDomainKnowledge.let { resultJson["filledByDomainKnowledge"] = it  }
    dateResult.successfullyParsed.let { resultJson["successfullyParsed"] = it  }

    return resultJson
  }

  private fun serializeSingleSideScanningResult(singleSideResult: BlinkCardSingleSideScanningResult): Map<String, Any> {
    val resultJson: MutableMap<String, Any> = mutableMapOf()
    singleSideResult.cardImage?.let { resultJson["cardImage"] = serializeCroppedImageResult(it) }
    singleSideResult.cardLivenessCheckResult.let {  resultJson["cardLivenessCheckResult"] = serializeCardLivenessCheckResult(it) }
    return resultJson
  }

  private fun serializeCardLivenessCheckResult(cardLivenessCheckResult: CardLivenessCheckResult): Map<String, Any> {
    val resultJson: MutableMap<String, Any> = mutableMapOf()
    cardLivenessCheckResult.photocopyCheckResult.let { resultJson["photocopyCheckResult"] = it.name.lowercase() }
    cardLivenessCheckResult.screenCheckResult.let { resultJson["screenCheckResult"] = it.name.lowercase() }
    cardLivenessCheckResult.cardHeldInHandCheckResult.let { resultJson["cardHeldInHandCheckResult"] = it.name.lowercase() }
    return resultJson
  }

  private fun serializeCroppedImageResult(croppedImageResult: CroppedImageResult): Map<String, String?> {
    return mapOf("image" to encodeBitmap(croppedImageResult.bitmap))
  }

  private fun encodeBitmap(bitmap: Bitmap?): String? {
    return bitmap?.let { bm ->
      val outputStream = ByteArrayOutputStream()
      outputStream.use { stream ->
        bm.compress(Bitmap.CompressFormat.JPEG, 95, stream)
        Base64.encodeToString(stream.toByteArray(), Base64.NO_WRAP)
      }
    }
  }
}

File: BlinkCard/ios/Serialization/BlinkCardSerializationUtils.swift
//
//  BlinkCardSerializationUtils.swift
//  BlinkCardReactNative
//
//  Created by Milan Parađina on 10.03.2026..
//

import Foundation
import UIKit
import BlinkCard

struct BlinkCardSerializationUtils {
    static func serializeBlinkCardScanningResult(_ result: BlinkCardScanningResult) -> String {
        var resultDict: [String: Any] = [
            "issuingNetwork": result.issuingNetwork,
            "cardAccounts": result.cardAccounts.map( {serializeCardAccountResult($0)} )
        ]
        
        if let iban = result.iban {
            resultDict["iban"] = iban
        }
        
        if let cardholderName = result.cardholderName {
            resultDict["cardholderName"] = cardholderName
        }
        
        resultDict["overallCardLivenessResult"] = result.overallCardLivenessResult.rawValue
        
        if let firstSideResult = result.firstSideResult {
            resultDict["firstSideResult"] = serializeSingleSideScanningResult(firstSideResult)
        }
        
        if let secondSideResult = result.secondSideResult {
            resultDict["secondSideResult"] = serializeSingleSideScanningResult(secondSideResult)
        }
        
        return encodeToJson(resultDict)
    }
    
    static func serializeCardAccountResult(_ cardAccountResult: CardAccountResult) -> [String: Any] {
        var cardAccountDict: [String: Any] = [
            "cardNumber": cardAccountResult.cardNumber,
            "cardNumberValid": cardAccountResult.cardNumberValid
        ]
        
        if let cardNumberPrefix = cardAccountResult.cardNumberPrefix {
            cardAccountDict["cardNumberPrefix"] = cardNumberPrefix
        }
        if let cvv = cardAccountResult.cvv {
            cardAccountDict["cvv"] = cvv
        }
        if let expiryDate = cardAccountResult.expiryDate {
            cardAccountDict["expiryDate"] = serializeDateResult(expiryDate)
        }
        if let fundingType = cardAccountResult.fundingType {
            cardAccountDict["fundingType"] = fundingType
        }
        if let issuerName = cardAccountResult.issuerName {
            cardAccountDict["issuerName"] = issuerName
        }
        if let issuerCountryCode = cardAccountResult.issuerCountryCode {
            cardAccountDict["issuerCountryCode"] = issuerCountryCode
        }
        if let issuerCountry = cardAccountResult.issuerCountry {
            cardAccountDict["issuerCountry"] = issuerCountry
        }
        if let cardCategory = cardAccountResult.cardCategory {
            cardAccountDict["cardCategory"] = cardCategory
        }
        
        return cardAccountDict
    }
    
    static func serializeSingleSideScanningResult(_ singleSideScanningResult: SingleSideScanningResult) -> [String: Any] {
        var singleSideScanningResultDict: [String: Any] = [:]
        
        if let cardImage = singleSideScanningResult.cardImage {
            singleSideScanningResultDict["cardImage"] = serializeCroppedImageResult(cardImage)
        }
        
        singleSideScanningResultDict["cardLivenessCheckResult"] = serializeCardLivenessCheckResult(singleSideScanningResult.cardLivenessCheckResult)
        
        return singleSideScanningResultDict
    }
    
    static func serializeCroppedImageResult(_ croppedImageResult: CroppedImageResult) -> [String: Any] {
        return ["image": encodeImage(croppedImageResult.uiImage)]
    }
    
    static func serializeDateResult(_ dateResult: DateResult) -> [String: Any] {
        var dateResultDict: [String: Any] = [
            "originalString" : dateResult.originalString,
            "filledByDomainKnowledge" : dateResult.filledByDomainKnowledge,
            "successfullyParsed" : dateResult.successfullyParsed
        ]
        if let day = dateResult.day {
            dateResultDict["day"] = day
        }
        if let month = dateResult.month {
            dateResultDict["month"] = month
        }
        if let year = dateResult.year {
            dateResultDict["year"] = year
        }
            
        return dateResultDict
    }
    
    static func serializeCardLivenessCheckResult(_ cardLivenessCheckResult: CardLivenessCheckResult) -> [String: Any] {
        return [
            "screenCheckResult": cardLivenessCheckResult.screenCheckResult.rawValue,
            "photocopyCheckResult": cardLivenessCheckResult.photocopyCheckResult.rawValue,
            "cardHeldInHandCheckResult": cardLivenessCheckResult.cardHeldInHandCheckResult.rawValue,
        ]
    }
    
    static func encodeToJson(_ dict: [String: Any]) -> String {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dict, options: [])
            let resultForJson = String(data: jsonData, encoding: .utf8)
            return resultForJson ?? ""
        } catch {
            return ""
        }
    }
    static func encodeImage(_ image: UIImage?) -> String? {
        return image?.jpegData(compressionQuality: 1.0)?.base64EncodedString()
    }
}
